apps/ios_rpc/tvmrpc/rpc_server.h

```diff
@@ -0,0 +1,318 @@
+  std::string GetPath(const std::string& file_name) { return base_ + file_name; }
+
+  void CleanUp() const {
+    CleanDir(base_);
+  }
+ private:
+  std::string base_;
+};
+
+
+/*!
+ * \brief RPCServer RPC Server class.
+ * \param host The hostname of the server, Default=0.0.0.0
+ * \param port The port of the RPC, Default=9090
+ * \param port_end The end search port of the RPC, Default=9099
+ * \param tracker The address of RPC tracker in host:port format e.g. 10.77.1.234:9190 Default=""
+ * \param key The key used to identify the device type in tracker. Default=""
+ * \param custom_addr Custom IP Address to Report to RPC Tracker. Default=""
+ */
+class RPCServer {
+ public:
+  /*!
+   * \brief Constructor.
+   */
+  RPCServer(std::string host, int port, int port_end, std::string tracker_addr, std::string key,
+            std::string custom_addr, std::string work_dir)
+      : host_(std::move(host)),
+        port_(port),
+        my_port_(0),
+        port_end_(port_end),
+        tracker_addr_(std::move(tracker_addr)),
+        key_(std::move(key)),
+        custom_addr_(std::move(custom_addr)),
+        work_dir_(std::move(work_dir)),
+        tracker_(tracker_addr_, key_, custom_addr_) {}
+
+  /*!
+   * \brief Destructor.
+   */
+  ~RPCServer() {
+    try {
+      // Free the resources
+      listen_sock_.Close();
+      tracker_.Close();
+    } catch (...) {
+    }
+  }
+
+  /*!
+   * \brief Start Creates the RPC listen process and execution.
+   */
+  void Start() {
+    listen_sock_.Create();
+    my_port_ = listen_sock_.TryBindHost(host_, port_, port_end_);
+    LOG(INFO) << "bind to " << host_ << ":" << my_port_;
+    listen_sock_.Listen(1);
+    continue_processing = true;
+    proc_ = std::future<void>(std::async(std::launch::async, &RPCServer::ListenLoopProc, this));
+  }
+  
+  void Stop() {
+    continue_processing = false;
+    tracker_.Close();
+  }
+    
+  void setCompletionCallbacks(std::function<void()> callback_start, std::function<void()> callback_stop) {
+    completion_callback_start_ = callback_start;
+    completion_callback_stop_ = callback_stop;
+  }
+
+ private:
+  /*!
+   * \brief ListenLoopProc The listen process.
+   */
+  void ListenLoopProc() {
+    
+    while (continue_processing) {
+      support::TCPSocket conn;
+      support::SockAddr addr("0.0.0.0", 0);
+      std::string opts;
+      try {
+        // step 1: setup tracker and report to tracker
+        tracker_.TryConnect();
+        if (completion_callback_start_)
+          completion_callback_start_();
+        // step 2: wait for in-coming connections
+        AcceptConnection(&tracker_, &conn, &addr, &opts);
+      } catch (const char* msg) {
+        LOG(WARNING) << "Socket exception: " << msg;
+        // close tracker resource
+        tracker_.Close();
+        continue;
+      } catch (const std::exception& e) {
+        // close tracker resource
+        tracker_.Close();
+        LOG(WARNING) << "Exception standard: " << e.what();
```

Comment:
what's "standard" mean? should this come before `tracker_.Close()` in case that throws an error?

Reply:
It was a copy-paste form neighbour "cpp_rpc" application. At last version of this patch I decided to rewrite this part of code in nonblocking style with Obj-C streams, RunLoop and other stuff. There is no more this exception line in iOS RPC project.   

https://github.com/apache/tvm/blob/1c251f50ee616507bdfd8866408e7acf9888cc3f/apps/cpp_rpc/rpc_server.cc#L160